### PR TITLE
Configuration: allow custom health check endpoint

### LINKED ISSUE #2938: Custom `path` for health check (eg. `/healthz`)
**Is your feature request related to a problem? Please describe.**
I want to be able to set a custom path for the health check endpoint

**Describe the solution you'd like**
```yaml
health_check:
  listen: 0.0.0.0:4000
  enabled: true
  path: /healthz
```

**Describe alternatives you've considered**
Update our Kubernetes configuration

**Additional context**
I'm migrating from @apollo/gateway to Apollo router. We should be able to deploy our new gateway (router) without changing the configuration of its corresponding service in kubernetes. We adopted a GCP standard for all our services, looking up the `/healthz` endpoint. Unfortunately Apollo router is hardcoding this endpoint to `/health` and therefore we cannot deploy with zero downtime.

### PR DESCRIPTION
**Pull request by aaronArinder**
Adds a configuration option for custom health check endpoints, using `/health` as default

Fixes #2938

---

**Checklist**

Complete the checklist (and note appropriate exceptions) before the PR is marked ready-for-review.

- [x] Changes are compatible[^1]
- [x] Documentation[^2] completed
- [x] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [x] Unit Tests
    - [ ] Integration Tests
    - [ ] Manual Tests

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]: It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]: Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]: Tick whichever testing boxes are applicable. If you are adding Manual Tests, please document the manual testing (extensively) in the Exceptions.

### CODE REVIEW

File: apollo-router/src/configuration/snapshots/apollo_router__configuration__tests__schema_generation.snap [hunk 1]
```diff
@@ -1557,7 +1557,8 @@ expression: "&schema"
       "description": "Health check configuration",
       "default": {
         "listen": "127.0.0.1:8088",
-        "enabled": true
+        "enabled": true,
+        "endpoint": "/health"
```
Comment (lennyburdette) [inline apollo-router/src/configuration/snapshots/apollo_router__configuration__tests__schema_generation.snap:1561]: Call it path to mirror the listen config? https://www.apollographql.com/docs/router/configuration/overview

**Comment by burn2delete** [comment]
Could we also get the helm charts updated to support the customizable field as well?

File: .changesets/config_aaron_allow_custom_healthcheck_endpoint.md [hunk 2]
```diff
@@ -0,0 +1,9 @@
+### Configuration: allow custom health check endpoint ([Issue #2938](https://github.com/apollographql/router/issues/2938))
+
+Adds a configuration option for custom health check endpoints, using `/health` as default
+
+
+<!-- start metadata -->
+---
```
Comment (Geal) [inline .changesets/config_aaron_allow_custom_healthcheck_endpoint.md:7]:
```suggestion
```

**Review by Geal — APPROVED** [review]

File: apollo-router/src/configuration/snapshots/apollo_router__configuration__tests__schema_generation.snap (hunk 1 above)
Comment (aaronArinder) [reply apollo-router/src/configuration/snapshots/apollo_router__configuration__tests__schema_generation.snap]: done!

**Comment by aaronArinder** [comment]
@burn2delete: I hadn't planned on it, and probably wouldn't do it in this PR. I might do it as a follow-up over the weekend or whenever I'm free next.

File: .changesets/config_aaron_allow_custom_healthcheck_endpoint.md (hunk 2 above)
Comment (bnjjj) [reply .changesets/config_aaron_allow_custom_healthcheck_endpoint.md:7]: Did you commit @aaronArinder ?

File: apollo-router/src/configuration/mod.rs [hunk 3]
```diff
@@ -1188,10 +1206,20 @@ impl HealthCheck {
...
     #[builder]
-    pub(crate) fn fake_new(listen: Option<ListenAddr>, enabled: Option<bool>) -> Self {
+    pub(crate) fn fake_new(
+        listen: Option<ListenAddr>,
+        enabled: Option<bool>,
+        path: Option<String>,
+    ) -> Self {
+        let mut path = path.unwrap_or_else(default_health_check_path);
+        if !path.starts_with('/') {
+            path = format!("/{path}").to_string();
```
Comment (bnjjj) [inline apollo-router/src/configuration/mod.rs:1216]:
```suggestion
            path = format!("/{path}");
```

File: docs/source/configuration/health-checks.mdx [hunk 4]
```diff
@@ -11,6 +11,23 @@ You can change this by setting `health_check`:
 health_check:
   listen: 127.0.0.1:8088
   enabled: true
+  path: /health
```
Comment (bnjjj) [inline docs/source/configuration/health-checks.mdx:14]:
```suggestion
  path: /health # Optional, default: /health
```

**Review by garypen — APPROVED** [review]

File: docs/source/configuration/health-checks.mdx (hunk 4 above)
Comment (aaronArinder) [reply docs/source/configuration/health-checks.mdx:14]: oh yeah, this is a good call, thanks

File: .changesets/config_aaron_allow_custom_healthcheck_endpoint.md (hunk 2 above)
Comment (aaronArinder) [reply .changesets/config_aaron_allow_custom_healthcheck_endpoint.md:7]: shoot, no, I thought I had but I didn't

Comment (aaronArinder) [reply .changesets/config_aaron_allow_custom_healthcheck_endpoint.md]: committed

**Review by bnjjj — APPROVED** [review]

File: docs/source/configuration/health-checks.mdx [hunk 5]
```diff
@@ -11,6 +11,23 @@ You can change this by setting `health_check`:
 health_check:
   listen: 127.0.0.1:8088
   enabled: true
+  path: /health # Optional, default: /health
+```
+
+Each option is configurable. For example, we might want to set our health check to `127.0.0.1:8090/healthz`:
```
Comment (shorgi) [inline docs/source/configuration/health-checks.mdx:17]:
```suggestion
Each option is configurable. For example, we can set our health check endpoint to `127.0.0.1:8090/healthz`:
```

File: docs/source/configuration/health-checks.mdx [hunk 6]
```diff
@@ -11,6 +11,23 @@ You can change this by setting `health_check`:
...
+
+```yaml title="router.yaml"
+health_check:
+  listen: 127.0.0.1:8090
+  enabled: true
+  path: /healthz
+```
+
+We can also disable the health check:
```
Comment (shorgi) [inline docs/source/configuration/health-checks.mdx:26]:
```suggestion
We can also disable the health check endpoint:
```

**Review by shorgi — COMMENTED** [review]
Suggested copyedits that use `health check endpoint` terminology

File: .changesets/config_aaron_allow_custom_healthcheck_endpoint.md [hunk 7]
```diff
@@ -0,0 +1,7 @@
+### Configuration: allow custom health check endpoint ([Issue #2938](https://github.com/apollographql/router/issues/2938))
+
+Adds a configuration option for custom health check endpoints, using `/health` as default
```
Comment (shorgi) [inline .changesets/config_aaron_allow_custom_healthcheck_endpoint.md:3]:
```suggestion
### Configuration: allow custom health check endpoint ([Issue #2938](https://github.com/apollographql/router/issues/2938))

Adds a configuration option for custom health check endpoints, `health_check.path`, with `/health` as the default value.
```

**Review by shorgi — APPROVED** [review]

**Pull request merged**